Gui: Refactor navigation animations (attempt 2)

Rexbas opened the pull request:
- Animations are played through an Animator
- Standard animations have a fixed duration and inherit from QVariantAnimation
- The animation duration can be changed in the preferences
- Fix animations when using the two arrows above the NaviCube
- Start an animation or start and wait for an animation
- Replaces standard camera animations (e.g. by selecting standard views or NaviCube) with FixedTimeAnimation
- Replace View3DInventorViewer's CameraAnimation with FixedTimeAnimation
- Replace OpenInventor navigation style infinite spin animation with SpinningAnimation
- Stops an active animation when a new animation is started
- Stops an active animation when the user starts dragging, panning or zooming
- Refactor reorientCamera so it can be used in animations
- Enable animations by default
- Fixes #9415 

This is the second attempt of #9446.

@wwmayer, double clicking the play button in the View Turntable from the tools menu should not crash anymore. And the minimum and maximum speed are faster. Can you confirm this?

Rexbas commented:
@wwmayer, have you had any time to check whether the problems and crash you described (https://github.com/FreeCAD/FreeCAD/pull/9446#issuecomment-1735503473) have been resolved? I think they are resolved but if not then I would like to take another look.

wwmayer commented:
> have you had any time to check whether the problems 

Unfortunately, not yet. I hope to have a look at it soon.

Rexbas commented:
I added some commits to simplify some things. I also fixed a problem with animations that combine rotation and translation where the final position was not correct.

wwmayer commented on src/Gui/View3DPy.cpp:
```diff
@@ -777,10 +777,8 @@ Py::Object View3DInventorPy::getCameraOrientation()
-    int steps = 20;
-    int ms = 30;
-    if (!PyArg_ParseTuple(args.ptr(), "|O!ii",&Base::PlacementPy::Type,&p,&steps,&ms))
+    PyObject* p = nullptr;
```
This change is problematic because it may break existing Python macros.

Rexbas replied:
Yes true, we could either keep the steps and ms parameters and ignore them so it does not break, or we could break it. I feel like it is better to clean it up and make any breaking changes before a 1.0 release. But let me know what you think.

wwmayer replied:
In the past it was possible to write this Python code:
```
view = Gui.ActiveDocument.ActiveView

# Manually move to the desired position and then call...
pos1 = view.viewPosition()
# Manually move to the desired position and then call...
pos2 = view.viewPosition()
# Manually move to the desired position and then call...
pos3 = view.viewPosition()

view.viewPosition(pos1, 20, 2000)
view.viewPosition(pos2, 50, 5000)
view.viewPosition(pos3, 10, 1000)
```

To achieve a similar effect now the code must be changed to:
```
view = Gui.ActiveDocument.ActiveView

# Manually move to the desired position and then call...
pos1 = view.viewPosition()
# Manually move to the desired position and then call...
pos2 = view.viewPosition()
# Manually move to the desired position and then call...
pos3 = view.viewPosition()

par = App.ParamGet("User parameter:BaseApp/Preferences/View")
par.SetInt("AnimationDuration", 2000)
view.viewPosition(pos1)
par.SetInt("AnimationDuration", 5000)
view.viewPosition(pos2)
par.SetInt("AnimationDuration", 1000)
view.viewPosition(pos3)
```
The downside is that this has side-effects because changing AnmimationDuration parameter affects other functions. Therefore I recommend the following changes:
* Revert this change
* Add an overloaded method `View3DInventorViewer::moveCameraTo(const SbRotation& orientation, const SbVec3f& position, int duration)`
* Use this function here and pass "ms" as third parameter.
* "steps" will still be used for the Python interface but not used any more
* In startAnimation() replace the bool with an int and remove the access to the user parameters

Rexbas replied:
Sounds like a good plan, working on it.

Rexbas replied:
I added a commit that makes `view.viewPosition()` work as previously. If the duration is not passed to `viewPosition()` then it will be -1 by default. I did not overload `View3DInventorViewer::moveCameraTo(...)`  but instead I added a duration parameter with default value -1. `View3DInventorViewer::startAnimation(...)` now also has a duration parameter with default value -1. `View3DInventorViewer::startAnimation(...)` checks if the duration is valid, meaning more or equal to 0. If it is not valid then the duration is overwritten with the "AnimationDuration" parameter. So, If the duration is not passed explicitly in one of the previously mentioned methods then the duration will default to "AnimationDuration".

I also did not replace the bool "wait" because for `viewPosition()` it is desired to wait until the animation is finished but when selecting a standard view or a face on the navicube, waiting cause some side-effects (I don´t remember what exactly).

wwmayer replied:
> I also did not replace the bool "wait" because for viewPosition() it is desired to wait until the animation is finished

OK.

wwmayer commented on src/Gui/View3DInventorViewer.cpp:
```diff
@@ -3117,18 +3082,49 @@ SbBool View3DInventorViewer::isAnimating() const
+    int duration = App::GetApplication()
+                       .GetParameterGroupByPath("User parameter:BaseApp/Preferences/View")
+                       ->GetInt("AnimationDuration", 250);
+
```
IMO it's a bad idea to make this function dependent on a user parameter because it will be more complicated to run consecutive calls of viewPosition() with different speeds. Then instead of a boolean argument and int could be used to increase flexibility.

Rexbas replied:
> it will be more complicated to run consecutive calls of viewPosition() with different speeds.

I guess 61e419b88ffac8cdb9f874363270327a0ab3ed3c also fixes this.

wwmayer replied:
> I guess 61e419b also fixed this.

It's OK.

wwmayer commented on src/Gui/NavigationStyle.cpp:
```diff
@@ -353,171 +323,71 @@ SbBool NavigationStyle::lookAtPoint(const SbVec2s screenpos)
     SbVec3f hitpoint;
     hitpoint = picked->getPoint();
-    lookAtPoint(hitpoint);
+    this->rotationCenterFound = false;
```
The public function `void lookAtPoint(const SbVec3f&)` shouldn't be simply removed. You can implement it as
```
void NavigationStyle::lookAtPoint(const SbVec3f& pos)
{
    this->rotationCenterFound = false;
    translateCamera(hitpoint - getFocalPoint());
}
```
and leave its call in `lookAtPoint(const SbVec2s screenpos)`

Rexbas replied:
Resolved in 26ea9614e64b923008f29cbb06d4b6f45efc0b37

wwmayer commented on src/Gui/View3DInventorViewer.cpp:
```diff
@@ -3117,18 +3082,49 @@ SbBool View3DInventorViewer::isAnimating() const
 void View3DInventorViewer::stopAnimating()
 {
-    navigation->stopAnimating();
+    navigation->getAnimator()->stop();
```
Have you ever heard of the [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter)? You should keep the method NavigationStyle::stopAnimation() and inside there call `getAnimator()->stop()`.

Rexbas replied:
Never heard of it until now. Resolved in f28e3325812b2e3ce9c024db367dde0f31108382

wwmayer commented on src/Gui/View3DInventorViewer.cpp:
```diff
@@ -3117,18 +3082,49 @@ SbBool View3DInventorViewer::isAnimating() const
 {
-    navigation->startAnimating(axis, velocity);
+    auto animation = std::make_shared<SpinningAnimation>(navigation, axis, velocity);
+    navigation->getAnimator()->start(animation);
```
Law of Demeter: Adjust NavigationStyle::startAnimating() to accept the SpinningAnimation and inside there call `getAnimator()->start(animation)`

wwmayer commented on src/Gui/View3DInventorViewer.cpp:
```diff
@@ -3117,18 +3082,49 @@ SbBool View3DInventorViewer::isAnimating() const
+    auto animation = std::make_shared<FixedTimeAnimation>(
+        navigation, orientation, rotationCenter, translation, duration);
+
+    if (wait) {
```
Law of Demeter

Rexbas replied:
Resolved in f28e3325812b2e3ce9c024db367dde0f31108382

wwmayer commented on src/Gui/View3DInventorViewer.cpp:
```diff
@@ -3117,18 +3082,49 @@ SbBool View3DInventorViewer::isAnimating() const
+        navigation->getAnimator()->startAndWait(animation);
+    }
+    else {
+        navigation->getAnimator()->start(animation);
```
Law of Demeter

Rexbas replied:
Resolved in f28e3325812b2e3ce9c024db367dde0f31108382

wwmayer commented on src/Gui/NavigationAnimation.cpp:
```diff
@@ -0,0 +1,167 @@
+/**
+ * @param finished True when the animation is finished, false when interrupted
+ */
+void SpinningAnimation::onStop(bool finished)
```
`finished` isn't used. Either use `Q_UNUSED(finished)` or declare it as `[[maybe_unused]]`

Rexbas replied:
Resolved in 2fec778d31a44790854e6ea224d0a436664d8096

wwmayer commented on src/Gui/NavigationAnimation.cpp:
```diff
@@ -0,0 +1,167 @@
+    update(value);
+}
+
+void NavigationAnimation::onStop(bool finished)
```
Unused parameter: Use `Q_UNUSED` or `[[maybe_unused]]`

Rexbas replied:
Resolved in 2fec778d31a44790854e6ea224d0a436664d8096

wwmayer commented:
With the PR there is still the issue that independent of the current camera location the duration to the new location is always fix which is weird for very small movements. I think we can leave this for now but should address it if there are more user complaints.

wwmayer commented:
> have you had any time to check whether the problems and crash you described

Yes, the crash is fixed and the behaviour of the speed slider is as expected. Thanks!

Rexbas commented:
> With the PR there is still the issue that independent of the current camera location the duration to the new location is always fix which is weird for very small movements. I think we can leave this for now but should address it if there are more user complaints.

If users start to complain then maybe I can add another type of animation with fixed velocity and let the user choose which type of animation to use.

There are also some other small improvements I have planned to make regarding navigation animations, but they will come later in small follow up PRs

Rexbas commented:
> Out of curiosity what exactly was the reason for this crash? https://github.com/FreeCAD/FreeCAD/pull/9446#issuecomment-1734316113

There was a segfault in the tests. The camera was nullptr but I am not sure why. The segfault was gone after I added a nullptr check in FixedTimeAnimation::update().

wwmayer commented:
> There was a segfault in the tests. The camera was nullptr but I am not sure why.

Ah, OK. Yes it can happen that the camera is a nullptr but I don't know the exact circumstances either.
